### DISCUSSION TITLE
mxnet convert to onnx: No conversion function registered for op type null yet.

### DISCUSSION THREAD

**Discussion by lookup1980** [apache/incubator-mxnet#15788] [Q&A]
## Description
I tried to convert mxnet mode to onnx, but got this error: No conversion function registered for op type null yet.

onnx version is:
```
>>> import onnx
>>> print(onnx.__version__)
1.5.0
```

The model and my code is:
```
sym = './resnet-18-symbol.json'
params = './resnet-18-0000.params'

onnx_file = './mxnet_exported.onnx'

input_shape = (1,3,512,512)

converted_model_path = onnx_mxnet.export_model(sym, params, [input_shape], np.float32, onnx_file)
```

## Environment info (Required)

----------Python Info----------
Version      : 3.7.3
Compiler     : GCC 7.3.0
Build        : ('default', 'Mar 27 2019 22:11:17')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 19.1.1
Directory    : /home/mgu/anaconda3/envs/mxnet-py37-cpu/lib/python3.7/site-packages/pip
----------MXNet Info-----------
Version      : 1.5.0
Directory    : /home/mgu/anaconda3/envs/mxnet-py37-cpu/lib/python3.7/site-packages/mxnet
Commit Hash   : 75a9e187d00a8b7ebc71412a02ed0e3ae489d91f
Library      : ['/home/mgu/anaconda3/envs/mxnet-py37-cpu/lib/python3.7/site-packages/mxnet/libmxnet.so']
Build features:
✖ CUDA
✖ CUDNN
✖ NCCL
✖ CUDA_RTC
✖ TENSORRT
✔ CPU_SSE
✔ CPU_SSE2
✔ CPU_SSE3
✔ CPU_SSE4_1
✔ CPU_SSE4_2
✖ CPU_SSE4A
✔ CPU_AVX
✖ CPU_AVX2
✖ OPENMP
✖ SSE
✔ F16C
✖ JEMALLOC
✖ BLAS_OPEN
✖ BLAS_ATLAS
✖ BLAS_MKL
✖ BLAS_APPLE
✔ LAPACK
✖ MKLDNN
✔ OPENCV
✖ CAFFE
✖ PROFILER
✔ DIST_KVSTORE
✖ CXX14
✖ INT64_TENSOR_SIZE
✔ SIGNAL_HANDLER
✖ DEBUG
----------System Info----------
Platform     : Linux-4.15.0-55-generic-x86_64-with-debian-stretch-sid
system       : Linux
node         : mgu-P520c
release      : 4.15.0-55-generic
version      : #60~16.04.2-Ubuntu SMP Thu Jul 4 09:03:09 UTC 2019
----------Hardware Info----------
machine      : x86_64
processor    : x86_64
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) W-2123 CPU @ 3.60GHz
Stepping:              4
CPU MHz:               1200.080
CPU max MHz:           3900.0000
CPU min MHz:           1200.0000
BogoMIPS:              7200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              8448K
NUMA node0 CPU(s):     0-7
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0061 sec, LOAD: 0.5550 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0462 sec, LOAD: 0.0659 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.0188 sec, LOAD: 0.1447 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0236 sec, LOAD: 0.1065 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0056 sec, LOAD: 0.3537 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0158 sec, LOAD: 0.0744 sec.

Package used (Python/R/Scala/Julia):
(I'm using ...)

For Scala user, please provide:
1. Java version: (`java -version`)
2. Maven version: (`mvn -version`)
3. Scala runtime if applicable: (`scala -version`)

For R user, please provide R `sessionInfo()`:

## Build info (Required if built from source)

Compiler (gcc/clang/mingw/visual studio):

MXNet commit hash:
(Paste the output of `git rev-parse HEAD` here.)

Build config:
(Paste the content of config.mk, or the build command.)

## Error Message:
(Paste the complete error message, including stack trace.)

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1.
2.

## What have you tried to solve it?

1.
2.

**Comment by vdantu** [comment]
@mxnet-label-bot add [onnx, question]
@vandanavk would you be able to answer this?

@lookup1980 Does your onnx_mxnet come from mx.contrib.onnx_mxnet?

**Comment by vandanavk** [comment]
@lookup1980 could you share your sym and params?
I tried with the resnet18 model in [GluonCV model zoo ](https://gluon-cv.mxnet.io/model_zoo/classification.html) and dint face any issues.

```
import mxnet as mx
import gluoncv
from gluoncv import model_zoo
from mxnet.contrib import onnx as onnx_mxnet
import numpy as np

net = model_zoo.get_model('resnet18_v2', pretrained=True)
net.hybridize()
x = np.zeros([1,3,224,244])
x = mx.nd.array(x)
net.forward(x)
net.export('resnet18_v2')

converted_model_path = onnx_mxnet.export_model('resnet18_v2-symbol.json', 'resnet18_v2-0000.params', [(1,3,224,224)], np.float32, 'test_resnet18.onnx')
```

**Comment by alexst07** [comment]
@vandanavk even your example gave me an error:

```
 File "conversion.py", line 14, in <module>
    converted_model_path = onnx_mxnet.export_model('resnet18_v2-symbol.json', 'resnet18_v2-0000.params', [(1,3,224,224)], np.float32, 'test_resnet18.onnx')
  File "/home/alex/anaconda3/lib/python3.7/site-packages/mxnet/contrib/onnx/mx2onnx/export_model.py", line 83, in export_model
    verbose=verbose)
  File "/home/alex/anaconda3/lib/python3.7/site-packages/mxnet/contrib/onnx/mx2onnx/export_onnx.py", line 312, in create_onnx_graph_proto
    checker.check_graph(graph)
  File "/home/alex/anaconda3/lib/python3.7/site-packages/onnx/checker.py", line 52, in checker
    proto.SerializeToString(), ctx)
onnx.onnx_cpp2py_export.checker.ValidationError: Unrecognized attribute: spatial for operator BatchNormalization

==> Context: Bad node spec: input: "data" input: "resnetv20_batchnorm0_gamma" input: "resnetv20_batchnorm0_beta" input: "resnetv20_batchnorm0_running_mean" input: "resnetv20_batchnorm0_running_var" output: "resnetv20_batchnorm0_fwd" name: "resnetv20_batchnorm0_fwd" op_type: "BatchNormalization" attribute { name: "epsilon" f: 1e-05 type: FLOAT } attribute { name: "momentum" f: 0.9 type: FLOAT } attribute { name: "spatial" i: 0 type: INT }
```

And yes, I had the same error of @lookup1980 before.

**Comment by vandanavk** [comment]
@alexst07 Please have a look at https://github.com/apache/incubator-mxnet/issues/14589#issuecomment-479057849 for information on the error with BatchNorm operator.

**Comment by Phxmark** [comment]
I face the same issus too.
And I found something in sym I used such as 
 ` {
      "op": "null", 
      "name": "data", 
      "inputs": []
    }, 
    {
      "op": "null", 
      "name": "bn_data_gamma", 
      "inputs": []
    }, 
    {
      "op": "null", 
      "name": "bn_data_beta", 
      "inputs": []
    }
`
how can I solve this

**Reply by mohamadHN93** [reply]
I encountered the same problem when converting mxnet to onnx
No conversion function registered for op type null yet
@vandanavk

**Comment by mohamadHN93** [comment]
I tried to convert a retinaface model to onnx but faced the same error. I would be grateful if you help me
import mxnet as mx
import numpy as np
from mxnet.contrib import onnx as onnx_mxnet
sym = 'R50-symbol.json'
params = 'R50-0000.params'
# Standard Imagenet input - 3 channels, 224*224
input_shape = (1,3,112,112)
# Path of the output file
onnx_file = 'mxnet_exported_resnet50.onnx'
converted_model_path = onnx_mxnet.export_model(sym, params, [input_shape], np.float32, onnx_file)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Users\mh.nakhaei\PycharmProjects\untitled\venv\lib\site-packages\mxnet\contrib\onnx\mx2onnx\export_model.py", line 79, in export_model
    verbose=verbose)
  File "C:\Users\mh.nakhaei\PycharmProjects\untitled\venv\lib\site-packages\mxnet\contrib\onnx\mx2onnx\export_onnx.py", line 234, in create_onnx_graph_proto
    index_lookup=index_lookup)
  File "C:\Users\mh.nakhaei\PycharmProjects\untitled\venv\lib\site-packages\mxnet\contrib\onnx\mx2onnx\export_onnx.py", line 86, in convert_layer
    raise AttributeError("No conversion function registered for op type %s yet." % op)
AttributeError: No conversion function registered for op type null yet.

**Reply by Zha0q1** [reply]
Hi MXNet 1.9 release is going to be out soon and it includes a major update to mx2onnx conversion
For early access you can refer to https://github.com/apache/incubator-mxnet/tree/v1.9.x/python/mxnet/onnx#installation to build a simple wheel for mx2onnx

**Reply by mohamadHN93** [reply]
Hi, I built it and added to my libraries, but still in conversion encountered the same error
AttributeError: No conversion function registered for op type null yet.

**Reply by chendongliang87** [reply]
same thing happened to me. any update?
